### PR TITLE
Add TuManhwas

### LINKED ISSUE #3222: TuManhwas
### Source name

TuManhwas

### Source link

https://tumanhwas.com/

### Source language

Spanish

### Other details

_No response_

### Acknowledgements

- [X] I have checked that the extension does not already exist by searching the [extension listing](https://keiyoushi.github.io/extensions/) and verified it does not appear there.
- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a meaningful title with the source name.
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by choppeh**
Closes #3222

Checklist:

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [ ] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [x] Have removed `web_hi_res_512.png` when adding a new extension

### CODE REVIEW

**Comment by bapeey** [comment]
Theme not seems like MangaThemesia, it has some of the same selectors and assets names but the structure is very different.

It seems like they don't even use Wordpress.

**Comment by choppeh** [comment]
> Theme not seems like MangaThemesia, it has some of the same selectors and assets names but the structure is very different.
> 
> It seems like they don't even use Wordpress.

I agree, I had to make so many changes that it was simpler not to use the theme

**Comment by vetleledaal** [comment]
> Theme not seems like MangaThemesia, 

My bad 🤐

Moving it out of the multisrc seems like the right choice

File: src/es/tumanhwas/src/eu/kanade/tachiyomi/extension/es/tumanhwas/TuManhwas.kt [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+            .build()
+        return GET(url, headers)
+    }
+
+    override fun searchMangaSelector() = popularMangaSelector()
+
+    override fun latestUpdatesFromElement(element: Element) = popularMangaFromElement(element).apply {
+        val chapterNumber = "-${url.substringAfterLast("-")}"
+        url = url.replace("news", "manga")
+            .removeSuffix(chapterNumber)
```
Comment (AwkwardPeak7) [inline src/es/tumanhwas/src/eu/kanade/tachiyomi/extension/es/tumanhwas/TuManhwas.kt]:
```suggestion
        url = url.replace("news", "manga")
            .substringBeforeLast('-')
```

File: src/es/tumanhwas/src/eu/kanade/tachiyomi/extension/es/tumanhwas/TuManhwas.kt [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    override fun latestUpdatesSelector() = ".bixbox.seriesearch:has(h1) a"
+
+    override fun mangaDetailsParse(document: Document) = SManga.create().apply {
+        document.selectFirst(".main-info")!!.let {
+            title = it.selectFirst("h1")!!.text()
+            thumbnail_url = it.selectFirst("img")?.imgSrc
+            description = it.selectFirst(".summary p")?.text()
+            genre = it.select(".genres-container a")
+                .map { it.text() }
+                .joinToString()
```
Comment (AwkwardPeak7) [inline src/es/tumanhwas/src/eu/kanade/tachiyomi/extension/es/tumanhwas/TuManhwas.kt]:
```suggestion
            genre = it.select(".genres-container a")
                .joinToString { it.text() }
```

File: src/es/tumanhwas/src/eu/kanade/tachiyomi/extension/es/tumanhwas/TuManhwas.kt [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+    override fun pageListParse(document: Document): List<Page> {
+        return document.select("#chapter_imgs img").mapIndexed { index, element ->
+            Page(index, document.location(), imageUrl = element.imgSrc)
+        }
+    }
+
+    override fun imageUrlParse(document: Document) = ""
+
+    private val Element.imgSrc: String get() = attr("abs:data-src").ifEmpty { attr("abs:src") }
```
Comment (AwkwardPeak7) [inline src/es/tumanhwas/src/eu/kanade/tachiyomi/extension/es/tumanhwas/TuManhwas.kt]:
```
private fun Element.imgAttr(): String {
    return when {
        hasAttr("data-src") -> absUrl("data-src")
        else -> absUrl("src")
    }
}
```
use something like this

File: src/es/tumanhwas/src/eu/kanade/tachiyomi/extension/es/tumanhwas/TuManhwas.kt [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+    private fun parseRelativeDate(date: String): Long {
+        val number = Regex("""(\d+)""").find(date)?.value?.toIntOrNull() ?: return 0
```
Comment (AwkwardPeak7) [inline src/es/tumanhwas/src/eu/kanade/tachiyomi/extension/es/tumanhwas/TuManhwas.kt]: pull out the regex

**Review by AwkwardPeak7 — CHANGES_REQUESTED** [review]

**Review by AwkwardPeak7 — APPROVED** [review]

**Pull request merged**